src/partition/repart.c

```diff
@@ -3972,32 +3970,33 @@ static int context_open_copy_block_paths(
                                                        "Copying from block device node is not permitted in --image=/--root= mode, refusing.");
 
                 } else if (p->copy_blocks_auto) {
+                        _cleanup_(sd_device_unrefp) sd_device *dev = NULL;
 
                         r = resolve_copy_blocks_auto(p->type_uuid, root, restrict_devno, &opened, &uuid);
                         if (r < 0)
                                 return r;
 
-                        source_fd = open(opened, O_RDONLY|O_CLOEXEC|O_NOCTTY);
+                        r = sd_device_new_from_path(&dev, opened);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to create sd-device object for device '%s': %m", opened);
+
+                        source_fd = sd_device_open(dev, O_RDONLY|O_CLOEXEC|O_NOCTTY);
```

Comment:
By the way, is the O_NOCTTY flag correct? Why? Maybe O_NONBLOCK?

Reply:
So, O_NOCTTY is probably something we always should specify if there's the tiniest of chances that what we are opening there ends up being a tty. I mean, if Linux had a nicer API it would be inverted, and be called O_CTTY because the behaviour it enables should be opt-out, not opt-in. (In fact some BSDs do just that.)

O_NONBLOCK only carries value on some specific devices, and fifos. It's a much unclearer situation.

I guess the rule should be:

1. Specify O_NOCTTY in all open() calls, except when O_DIRECTORY (because if something is a dir, it definitely isn't a tty), O_PATH (because an inode fd definitely is not a tty), O_CREAT|O_EXCL (a new file cannot be a tty) or O_TMPFILE (same) is already specified.

2. Specify O_NONBLOCK whenever we expect to open a regular file only. Because not specifying O_NONBLOCK means we might otherwise hang in the open() call.

(I guess one of those days we should add a wrapper open_regular() or so that safely opens a regular file and safely refuses weird stuff like fifos or special devices, and will not block on them in any cases)

Reply:
Thanks.